### PR TITLE
Receive and save the script to disk

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Previously only the filepath to the script was sent to the server, as it is assumed all the nodes share a filesystem. This assumption maybe invalid and has been removed.

Instead the script must be sent with the job when it is created. It is than saved to disk so it can be retrieved latter. This required various changes to the internals to remove the old `script` method and replace it with `read_script`. The over-arching design principle is the `script` should not be cached on a `Job` object as it will remain in ram. Instead it has been written to disk asap and the read from disk when required.

The work flow for when the script is deleted could be improved. ATM it is deleted by scheduler

Finally because we are working with files, it is handy to fake the filesystem to prevent the excessive build up of files. This needs to be activated manually to prevent any surprises.

PS Validation errors now return 422 instead of 500

### CODE REVIEW

File: lib/flight_scheduler/event_processor.rb [hunk 1]
```diff
@@ -50,6 +50,7 @@ def cancel_job(job)
     Async.logger.info("Canceling job #{job.id}")
     return unless %w(PENDING RUNNING).include?(job.state)
 
+    # NOTE: Is the above and below state lines correct? Needs review
```
Comment (WilliamMcCumstie) [inline lib/flight_scheduler/event_processor.rb]: I came across this section whilst refactoring but couldn't determine what it was meant to do. Two things:

1. Is line 54 meant to be `job.state = CANCELLED`? ATM it is doing an equality check but not using the result.
2. What is the purpose of line 55 `return if <pending>` considering we do a similar check above?

Comment (benarmston) [reply lib/flight_scheduler/event_processor.rb]: Job cancellation has seen some refactoring in #8.  FWIW: line 54 should have been an assignment.  Line 51, prevents cancelling completed jobs and line 55 prevents attempting to kill processes for pending jobs (obviously the processes don't exist).

File: app.rb [hunk 2]
```diff
@@ -190,8 +195,11 @@ def find(id)
       end
 
       def validate!
-        if resource.validate!
+        if @created && resource.validate!
```
Comment (benarmston) [inline app.rb:200]: What's the purpose of `@created` here?  I thought `validate!` was only called after the `create` action.  Is it being called at other times too?

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb [hunk 3]
```diff
@@ -42,8 +42,13 @@ def add_job(job)
   end
 
   # Remove a single job from the queue.
+  # TODO: Having 'job.cleanup' is less than ideal, however it is the only place
+  # that is guaranteed to be called regardless what happens to the job
+  # e.g. Cancelled, fails, exits normally
+  # consider refactoring
   def remove_job(job)
     @queue.delete(job)
+    job.cleanup
```
Comment (benarmston) [inline lib/flight_scheduler/schedulers/fifo_scheduler.rb:51]: This is fine for now.  But removing a job from the scheduler and cleaning up a job are two different concerns.  Once this branch and the array tasks branch are both merged, we can look at a better place for this to live.

**Review by benarmston — COMMENTED** [review]
I've left a comment in-line below.

File: app.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply app.rb:200]: ATM you are correct that `validate!` only gets called during `create`, however in general this is not the case. Adding a `update` (possible `pluck`, `prune`, `graft`, etc) route will also trigger the `validate!` and break the scheduler.

By using a `@create` variable it explicitly prevents jobs being re-added in the future.

tl;dr It is not required beyond best practice

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/schedulers/fifo_scheduler.rb:51]: :+1:

File: lib/flight_scheduler/event_processor.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/event_processor.rb]: Removed the comment in light of your changes:
https://github.com/openflighthpc/flight-scheduler-controller/pull/7/commits/06e21a3335ad12454e58b36fdb028b60f5378fd9

File: app.rb (hunk 2 above)
Comment (benarmston) [reply app.rb:200]: Fair enough.

**Comment by benarmston** [comment]
LGTM.
